File: src/UnwindMC/Generation/Ast/AstBuilder.fs
module rec AstBuilder

open System.Collections.Generic
open Ast
open IL
open FlowAnalyzer
open Type

type Block = Block<ILOperand>
type Instruction = ILInstruction<ILOperand>

type private T = {
    blocks: IReadOnlyList<Block>
    variableTypes: IReadOnlyList<DataType>
    variableNames: Dictionary<int, string>
    parameterNames: Dictionary<int, string>
    localNames: Dictionary<int, string>
    types: Dictionary<string, DataType>
    mutable nextVariableNameIdx: int
}

type Variable = {
    name: string
    type_: DataType
}

type Function = {
    name: string
    parameters: IReadOnlyList<Variable>
    locals: IReadOnlyList<Variable>
    variables: IReadOnlyList<Variable>
    body: IReadOnlyList<Statement>
}

let buildAst (name: string) (blocks: IReadOnlyList<Block>) (parameterTypes: IReadOnlyList<DataType>)
    (localTypes: IReadOnlyList<DataType>) (variableTypes: IReadOnlyList<DataType>): Function =
    let t = {
        blocks = blocks
        variableTypes = variableTypes
        variableNames = new Dictionary<int, string>()
        parameterNames = new Dictionary<int, string>()
        localNames = new Dictionary<int, string>()
        types = new Dictionary<string, DataType>()
        nextVariableNameIdx = 0
    }
    let mutable offset = 0
    for index in [0 .. parameterTypes.Count - 1] do
        let name = "arg" + string(index)
        t.parameterNames.[offset] <- name
        t.types.[name] <- parameterTypes.[index]
        offset <- offset + sizeOf parameterTypes.[index]
    offset <- -Constants.RegisterSize * 2 // TODO: this will not always be correct
    for index in [0 .. localTypes.Count - 1] do
        offset <- offset - sizeOf localTypes.[index]
        let name = "loc" + string(index)
        t.localNames.[offset] <- name
        t.types.[name] <- localTypes.[index]
    let body =
        buildScope t t.blocks
        |> ROL.map (runTransformations t)
    {
        name = name
        parameters = t.parameterNames |> Seq.map (fun p -> { name = p.Value; type_ = t.types.[p.Value] }) |> Seq.toArray
        locals = t.localNames |> Seq.map (fun p -> { name = p.Value; type_ = t.types.[p.Value] }) |> Seq.toArray
        variables = t.variableNames |> Seq.map (fun p -> { name = p.Value; type_ = t.variableTypes.[p.Key] }) |> Seq.toArray
        body = body
    }

let private runTransformations (t: T) (ast: Statement): Statement =
    [
        FixupPointerArithmetics.transformer t.types
        FixupZeroAssignment.transformer
    ] |> Seq.fold (fun a t -> Transformer.transform t a) ast

let private buildScope (t: T) (blocks: IReadOnlyList<Block>): IReadOnlyList<Statement> =
    let statements = new List<Statement>()
    for block in blocks do
        match block with
        | SequentialBlock { instructions = is } ->
            for instr in is |> Seq.filter (function Nop -> false | _ -> true) do
                statements.Add(buildStatement t instr);
        | WhileBlock { condition = c; body = cs } ->
            statements.Add(buildWhile t c cs)
        | DoWhileBlock { condition = c; body = cs } ->
            statements.Add(buildDoWhile t c cs)
        | ForBlock { condition = c; modifier = m; body = b } ->
            statements.Add(buildFor t c m b)
        | ConditionalBlock { condition = c; trueBranch = tb; falseBranch = fb } ->
            statements.Add(buildIfThenElse t c tb fb)
    statements :> IReadOnlyList<Statement>

let private buildWhile (t: T) (condition: IReadOnlyList<Instruction>) (children: IReadOnlyList<Block>): Statement =
    While (buildCondition t condition, buildScope t children)

let private buildDoWhile (t: T) (condition: IReadOnlyList<Instruction>) (children: IReadOnlyList<Block>): Statement =
    DoWhile (buildScope t children, buildCondition t condition)

let private buildFor (t: T) (condition: IReadOnlyList<Instruction>) (modifier: IReadOnlyList<Instruction>) (body: IReadOnlyList<Block>): Statement =
    For (buildCondition t condition, buildScope t [| (SequentialBlock { instructions = modifier }) |], buildScope t body)

let private buildIfThenElse (t: T) (condition: IReadOnlyList<Instruction>) (trueBranch: IReadOnlyList<Block>) (falseBranch: IReadOnlyList<Block>): Statement =
    IfThenElse (buildCondition t condition, buildScope t trueBranch, buildScope t falseBranch)

let private buildCondition (t: T) (condition: IReadOnlyList<Instruction>): Expression =
    match (condition |> Seq.toList) with
    | (Compare compare) :: (Branch branch) :: [] ->
        buildBinaryOperator t (getBinaryOperator branch.type_) compare
    | (Assign assign) :: (Compare compare) :: (Branch branch) :: [] when assign.left = compare.left ->
        // TODO: this is a special case where we can inline the assigned variable, remove this
        buildBinaryOperator t (getBinaryOperator branch.type_) { compare with left = assign.right }
    | _ -> failwith "Instruction is not a valid statement"

let private buildStatement (t: T) (instr: Instruction): Statement =
    match instr with
    | Binary (op, binary) ->
        Assignment (buildVar t binary.left binary.leftId, buildBinaryOperator t (convertBinaryOperator op) binary)
    | Unary (op, unary) ->
        Assignment (buildVar t unary.operand unary.operandId, buildUnaryOperator t (convertUnaryOperator op) unary)
    | Assign binary ->
        Assignment (buildVar t binary.left binary.leftId, buildExpression t binary.right binary.rightId)
    | Call unary ->
        FunctionCall (buildExpression t unary.operand unary.operandId)
    | Return unary ->
        Statement.Return (if unary.operandId = -1 then Option.None else Some(buildVar t unary.operand unary.operandId))
    | Continue ->
        Statement.Continue
    | Break ->
        Statement.Break
    | Nop
    | Compare _
    | Branch _ ->
        impossible

let private buildExpression (t: T) (op: ILOperand) (id: int): Expression =
    match op with
    | ILOperand.Pointer _ -> Dereference (VarRef (Var (getVarName t id)))
    | Register _ -> VarRef (Var (getVarName t id))
    | Argument offset -> VarRef (Var t.parameterNames.[offset])
    | Local offset -> VarRef (Var t.localNames.[offset])
    | Value value -> Expression.Value (value)
    | NoOperand -> impossible

let private getBinaryOperator (condition: BranchType): Operator =
    match condition with
    | Equal -> Operator.Equal
    | NotEqual -> Operator.NotEqual
    | Less -> Operator.Less
    | LessOrEqual -> Operator.LessOrEqual
    | GreaterOrEqual -> Operator.GreaterOrEqual
    | Greater -> Operator.Greater
    | Unconditional -> impossible

let private convertBinaryOperator (op: ILBinaryOperator): Operator =
    match op with
    | Add -> Operator.Add
    | And -> Operator.And
    | Divide -> Operator.Divide
    | Multiply -> Operator.Multiply
    | Or -> Operator.Or
    | ShiftLeft -> Operator.ShiftLeft
    | ShiftRight -> Operator.ShiftRight
    | Subtract -> Operator.Subtract
    | Xor -> Operator.Xor

let private convertUnaryOperator (op: ILUnaryOperator): Operator =
    match op with
    | Negate -> Operator.Negate
    | Not -> Operator.Not

let private buildBinaryOperator (t: T) (op: Operator) (instr: BinaryInstruction<ILOperand>): Expression =
    Expression.Binary (op, buildExpression t instr.left instr.leftId, buildExpression t instr.right instr.rightId)

let private buildUnaryOperator (t: T) (op: Operator) (instr: UnaryInstruction<ILOperand>): Expression =
    Expression.Unary (op, buildExpression t instr.operand instr.operandId)

let private buildVar (t: T) (op: ILOperand) (id: int): Var =
    match op with
    | Register _ -> Var (getVarName t id)
    | Argument offset -> Var t.parameterNames.[offset]
    | Local offset -> Var t.localNames.[offset]
    | _ -> notSupported

let private getVarName (t: T) (id: int): string =
    if id = -1 then
        failwith "Invalid id"
    else
        let hasValue, name = t.variableNames.TryGetValue(id)
        if hasValue then
            name
        else
            let name = "var" + string(t.nextVariableNameIdx)
            t.nextVariableNameIdx <- t.nextVariableNameIdx + 1
            t.variableNames.[id] <- name
            t.types.[name] <- t.variableTypes.[id]
            name

File: src/ast_builder.rs
use std::collections::HashMap;

use ast::*;
use common::REGISTER_SIZE;
use flow_analyzer::*;
use il::*;
use transformer::*;
use transformer_fixup_pointer_arithmetics;
use transformer_fixup_zero_assignment;
use type_resolver::*;

type ResolvedBlock = Block<ResolvedOperand>;
type Instruction = ILInstruction<ResolvedOperand>;

pub struct AstBuilder<'a> {
    variable_types: &'a Vec<DataType>,
    variable_names: HashMap<i32, String>,
    parameter_names: HashMap<i32, String>,
    local_names: HashMap<i32, String>,
    types: HashMap<String, DataType>,
    next_var_name_idx: i32,
}

#[derive(Debug)]
pub struct Variable {
    pub name: String,
    pub type_: DataType,
}

#[derive(Debug)]
pub struct Function {
    pub name: String,
    pub parameters: Vec<Variable>,
    pub locals: Vec<Variable>,
    pub variables: Vec<Variable>,
    pub body: Vec<Statement>,
}

impl<'a> AstBuilder<'a> {
    pub fn build_ast(name: String, blocks: &Vec<ResolvedBlock>, types: &Result) -> Function {
        let mut builder = AstBuilder {
            variable_types: &types.variable_types,
            variable_names: HashMap::new(),
            parameter_names: HashMap::new(),
            local_names: HashMap::new(),
            types: HashMap::new(),
            next_var_name_idx: 0,
        };

        let mut offset = 0;
        for index in 0..types.parameter_types.len() {
            let name = String::from("arg") + &index.to_string();
            builder.parameter_names.insert(offset, name.clone());
            builder
                .types
                .insert(name.clone(), types.parameter_types[index].clone());
            offset += types.parameter_types[index].size() as i32;
        }
        offset -= REGISTER_SIZE as i32 * 2; // TODO: this will not always be correct
        for index in 0..types.local_types.len() {
            offset -= types.local_types[index].size() as i32;
            let name = String::from("loc") + &index.to_string();
            builder.local_names.insert(offset, name.clone());
            builder
                .types
                .insert(name.clone(), types.local_types[index].clone());
        }

        let mut body = builder.build_scope(blocks);

        let transformers: Vec<Box<dyn Transformer>> = vec![
            Box::new(transformer_fixup_pointer_arithmetics::T::new(
                &builder.types,
            )),
            Box::new(transformer_fixup_zero_assignment::T::new()),
        ];

        body = transformers.into_iter().fold(body, |b, mut t| {
            b.into_iter().map(|s| t.transform_statement(s)).collect()
        });

        Function {
            name: name,
            parameters: builder
                .parameter_names
                .values()
                .map(|v| Variable {
                    name: v.clone(),
                    type_: builder.types.get(v).unwrap().clone(),
                }).collect(),
            locals: builder
                .local_names
                .values()
                .map(|v| Variable {
                    name: v.clone(),
                    type_: builder.types.get(v).unwrap().clone(),
                }).collect(),
            variables: builder
                .variable_names
                .values()
                .map(|v| Variable {
                    name: v.clone(),
                    type_: builder.types.get(v).unwrap().clone(),
                }).collect(),
            body: body,
        }
    }

    fn build_scope(&mut self, blocks: &Vec<ResolvedBlock>) -> Vec<Statement> {
        let mut statements = vec![];
        for block in blocks.iter() {
            match block {
                Block::SequentialBlock(block) => {
                    for insn in block.instructions.iter() {
                        match insn {
                            ILInstruction::Nop => (),
                            insn => statements.push(self.build_statement(insn)),
                        }
                    }
                }
                Block::WhileBlock(block) => {
                    statements.push(self.build_while(&block.condition, &block.body));
                }
                Block::DoWhileBlock(block) => {
                    statements.push(self.build_do_while(&block.condition, &block.body));
                }
                Block::ForBlock(block) => {
                    statements.push(self.build_for(&block.condition, &block.modifier, &block.body));
                }
                Block::ConditionalBlock(block) => {
                    statements.push(self.build_if_then_else(
                        &block.condition,
                        &block.true_branch,
                        &block.false_branch,
                    ));
                }
            }
        }
        statements
    }

    fn build_while(
        &mut self,
        condition: &Vec<Instruction>,
        body: &Vec<ResolvedBlock>,
    ) -> Statement {
        Statement::While(self.build_condition(condition), self.build_scope(body))
    }

    fn build_do_while(
        &mut self,
        condition: &Vec<Instruction>,
        body: &Vec<ResolvedBlock>,
    ) -> Statement {
        Statement::DoWhile(self.build_scope(body), self.build_condition(condition))
    }

    fn build_for(
        &mut self,
        condition: &Vec<Instruction>,
        modifier: &Vec<Instruction>,
        body: &Vec<ResolvedBlock>,
    ) -> Statement {
        Statement::For(
            self.build_condition(condition),
            self.build_scope(&vec![Block::SequentialBlock(SequentialBlock {
                instructions: modifier.clone(),
            })]),
            self.build_scope(body),
        )
    }

    fn build_if_then_else(
        &mut self,
        condition: &Vec<Instruction>,
        true_branch: &Vec<ResolvedBlock>,
        false_branch: &Vec<ResolvedBlock>,
    ) -> Statement {
        Statement::IfThenElse(
            self.build_condition(condition),
            self.build_scope(true_branch),
            self.build_scope(false_branch),
        )
    }

    fn build_condition(&mut self, condition: &Vec<Instruction>) -> Expression {
        match &condition[..] {
            [ILInstruction::Compare(compare), ILInstruction::Branch(branch)] => {
                self.build_binary_operator(AstBuilder::get_binary_operator(&branch.type_), &compare)
            }
            [ILInstruction::Assign(assign), ILInstruction::Compare(compare), ILInstruction::Branch(branch)] =>
            {
                // TODO: this is a special case where we can inline the assigned variable, remove this
                let mut compare = compare.clone();
                compare.left = assign.right;
                self.build_binary_operator(AstBuilder::get_binary_operator(&branch.type_), &compare)
            }
            _ => panic!("Instruction is not a valid statement"),
        }
    }

    fn build_statement(&mut self, insn: &Instruction) -> Statement {
        match insn {
            ILInstruction::Binary(op, binary) => Statement::Assignment(
                self.build_var(&binary.left),
                self.build_binary_operator(AstBuilder::convert_binary_operator(op), binary),
            ),
            ILInstruction::Unary(op, unary) => Statement::Assignment(
                self.build_var(&unary.operand),
                self.build_unary_operator(AstBuilder::convert_unary_operator(op), unary),
            ),
            ILInstruction::Assign(binary) => Statement::Assignment(
                self.build_var(&binary.left),
                self.build_expression(&binary.right),
            ),
            ILInstruction::Call(unary) => {
                Statement::FunctionCall(self.build_expression(&unary.operand))
            }
            ILInstruction::Return(unary) => {
                let (_, id) = unary.operand;
                Statement::Return(id.map(|_| self.build_var(&unary.operand)))
            }
            ILInstruction::Continue => Statement::Continue,
            ILInstruction::Break => Statement::Break,
            ILInstruction::Nop | ILInstruction::Compare(_) | ILInstruction::Branch(_) => {
                panic!("impossble")
            }
        }
    }

    fn get_binary_operator(condition: &BranchType) -> Operator {
        match condition {
            BranchType::Equal => Operator::Equal,
            BranchType::NotEqual => Operator::NotEqual,
            BranchType::Less => Operator::Less,
            BranchType::LessOrEqual => Operator::LessOrEqual,
            BranchType::Greater => Operator::Greater,
            BranchType::GreaterOrEqual => Operator::GreaterOrEqual,
            BranchType::Unconditional => panic!("impossible"),
        }
    }

    fn convert_binary_operator(op: &ILBinaryOperator) -> Operator {
        match op {
            ILBinaryOperator::Add => Operator::Add,
            ILBinaryOperator::And => Operator::And,
            ILBinaryOperator::Divide => Operator::Divide,
            ILBinaryOperator::Multiply => Operator::Multiply,
            ILBinaryOperator::Or => Operator::Or,
            ILBinaryOperator::ShiftLeft => Operator::ShiftLeft,
            ILBinaryOperator::ShiftRight => Operator::ShiftRight,
            ILBinaryOperator::Subtract => Operator::Subtract,
            ILBinaryOperator::Xor => Operator::Xor,
        }
    }

    fn convert_unary_operator(op: &ILUnaryOperator) -> Operator {
        match op {
            ILUnaryOperator::Negate => Operator::Negate,
            ILUnaryOperator::Not => Operator::Not,
        }
    }

    fn build_binary_operator(
        &mut self,
        op: Operator,
        insn: &BinaryInstruction<ResolvedOperand>,
    ) -> Expression {
        Expression::Binary(
            op,
            Box::new(self.build_expression(&insn.left)),
            Box::new(self.build_expression(&insn.right)),
        )
    }

    fn build_unary_operator(
        &mut self,
        op: Operator,
        insn: &UnaryInstruction<ResolvedOperand>,
    ) -> Expression {
        Expression::Unary(op, Box::new(self.build_expression(&insn.operand)))
    }

    fn build_expression(&mut self, op: &ResolvedOperand) -> Expression {
        let (op, id) = op;
        use ast::Expression::*;
        use il::ILOperand::*;
        match op {
            Pointer(_, _) => Dereference(Box::new(VarRef(Var::Var(self.get_var_name(id))))),
            Register(_) => VarRef(Var::Var(self.get_var_name(id))),
            Argument(offset) => VarRef(Var::Var(self.parameter_names.get(offset).unwrap().clone())),
            Local(offset) => VarRef(Var::Var(self.local_names.get(offset).unwrap().clone())),
            ILOperand::Value(value) => Expression::Value(*value),
        }
    }

    fn build_var(&mut self, op: &ResolvedOperand) -> Var {
        let (op, id) = op;
        use il::ILOperand::*;
        match op {
            Register(_) => Var::Var(self.get_var_name(id)),
            Argument(offset) => Var::Var(self.parameter_names.get(offset).unwrap().clone()),
            Local(offset) => Var::Var(self.local_names.get(offset).unwrap().clone()),
            _ => panic!("Not supported"),
        }
    }

    fn get_var_name(&mut self, id: &Option<i32>) -> String {
        match id {
            Some(id) => {
                let name = self.variable_names.get(&id);
                match name {
                    Some(name) => name.clone(),
                    None => {
                        let mut name = String::from("var");
                        name += &self.next_var_name_idx.to_string();
                        self.next_var_name_idx += 1;
                        self.variable_names.insert(*id, name.clone());
                        self.types
                            .insert(name.clone(), self.variable_types[*id as usize].clone());
                        name
                    }
                }
            }
            None => panic!("Invalid id"),
        }
    }
}
